File: lang/cpp/benchmarks/ffi_benchmark.cpp
/**
 * FFI Performance Benchmark for DBX C++ Bindings
 * 
 * Measures INSERT, GET, and SCAN performance to quantify FFI overhead
 * compared to Rust Core benchmarks.
 * 
 * Build: cmake --build build --target dbx_ffi_benchmark
 * Run: ./build/benchmarks/dbx_ffi_benchmark
 */

#include <benchmark/benchmark.h>
#include <string>
#include <vector>
#include "dbx.h"

constexpr int NUM_ENTRIES = 10'000;

struct TestData {
    std::vector<std::pair<std::string, std::string>> data;
    
    TestData() {
        data.reserve(NUM_ENTRIES);
        for (int i = 0; i < NUM_ENTRIES; i++) {
            char key[32], value[64];
            snprintf(key, sizeof(key), "key_%08d", i);
            snprintf(value, sizeof(value), "value_%08d_data", i);
            data.emplace_back(key, value);
        }
    }
};

static void BM_DBX_Insert(benchmark::State& state) {
    TestData test_data;
    
    for (auto _ : state) {
        auto db = dbx::Database::open_in_memory();
        for (const auto& [key, value] : test_data.data) {
            db.insert("bench", key, value);
        }
    }
}
BENCHMARK(BM_DBX_Insert);

static void BM_DBX_Get(benchmark::State& state) {
    TestData test_data;
    
    // Setup: Insert data first
    auto db = dbx::Database::open_in_memory();
    for (const auto& [key, value] : test_data.data) {
        db.insert("bench", key, value);
    }
    
    for (auto _ : state) {
        for (const auto& [key, _] : test_data.data) {
            db.get("bench", key);
        }
    }
}
BENCHMARK(BM_DBX_Get);

static void BM_DBX_Scan(benchmark::State& state) {
    TestData test_data;
    
    // Setup: Insert data first
    auto db = dbx::Database::open_in_memory();
    for (const auto& [key, value] : test_data.data) {
        db.insert("bench", key, value);
    }
    
    for (auto _ : state) {
        db.scan("bench");
    }
}
BENCHMARK(BM_DBX_Scan);

BENCHMARK_MAIN();

File: packages/nodejs/benchmarks/ffi_benchmark.js
/**
 * FFI Performance Benchmark for DBX Node.js Bindings
 * 
 * Measures INSERT, GET, and SCAN performance to quantify FFI overhead
 * compared to Rust Core benchmarks.
 * 
 * Run with: node benchmarks/ffi_benchmark.js
 */

const Benchmark = require('benchmark');
const { Database } = require('../');

const NUM_ENTRIES = 10_000;

function generateTestData() {
    const data = [];
    for (let i = 0; i < NUM_ENTRIES; i++) {
        const key = Buffer.from(`key_${i.toString().padStart(8, '0')}`);
        const value = Buffer.from(`value_${i.toString().padStart(8, '0')}_data`);
        data.push([key, value]);
    }
    return data;
}

// Prepare data for GET and SCAN benchmarks
let getDb, scanDb, testData;

const suite = new Benchmark.Suite();

suite
    .add('DBX INSERT 10k', function () {
        const db = Database.openInMemory();
        const data = generateTestData();
        for (const [key, value] of data) {
            db.insert('bench', key, value);
        }
    })
    .add('DBX GET 10k', {
        onStart: function () {
            getDb = Database.openInMemory();
            testData = generateTestData();
            for (const [key, value] of testData) {
                getDb.insert('bench', key, value);
            }
        },
        fn: function () {
            for (const [key] of testData) {
                getDb.get('bench', key);
            }
        }
    })
    .add('DBX SCAN 10k', {
        onStart: function () {
            scanDb = Database.openInMemory();
            const data = generateTestData();
            for (const [key, value] of data) {
                scanDb.insert('bench', key, value);
            }
        },
        fn: function () {
            scanDb.scan('bench');
        }
    })
    .on('cycle', function (event) {
        console.log(String(event.target));
        const timeMs = event.target.stats.mean * 1000;
        console.log(`  → ${timeMs.toFixed(2)}ms per operation\n`);
    })
    .on('complete', function () {
        console.log('\n=== Benchmark Complete ===');
    })
    .run({ 'async': false });
